src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/ServerConfigurationFactory.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+[Export, Shared]
+internal class ServerConfigurationFactory
+{
+    private ServerConfiguration? _serverConfiguration;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public ServerConfigurationFactory()
+    {
+    }
+
+    [Export(typeof(ServerConfiguration))]
+    public ServerConfiguration ServerConfiguration => _serverConfiguration ?? throw new InvalidOperationException($"{nameof(ServerConfiguration)} has not been initialized");
+
+    public void InitializeConfiguration(ServerConfiguration serverConfiguration) => _serverConfiguration = serverConfiguration;
```

Comment:
any concern about double initialization?  consider throwing if this happens more than once.